File: simplified-books-core-ng/src/main/java/org/nypl/simplified/books/accounts/AccountID.java
package org.nypl.simplified.books.accounts;

import com.google.auto.value.AutoValue;
import com.io7m.jnull.NullCheck;

import java.io.Serializable;

/**
 * A unique account identifier.
 */

@AutoValue
public abstract class AccountID implements Comparable<AccountID>, Serializable {

  AccountID() {

  }

  /**
   * Construct a account identifier.
   *
   * @param id A non-negative integer
   */

  public static AccountID create(final int id)
  {
    if (id < 0) {
      throw new IllegalArgumentException("Account identifiers must be non-negative");
    }
    return new AutoValue_AccountID(id);
  }

  /**
   * @return The raw identifier value
   */

  public abstract int id();

  @Override
  public int compareTo(final AccountID other) {
    return Integer.compare(this.id(), NullCheck.notNull(other, "Other").id());
  }
}

File: simplified-accounts-api/src/main/java/org/nypl/simplified/accounts/api/AccountID.kt
package org.nypl.simplified.accounts.api

import java.io.Serializable
import java.util.UUID

/**
 * A unique identifier for an account.
 */

data class AccountID(
  val uuid: UUID
) : Serializable, Comparable<AccountID> {

  override fun compareTo(other: AccountID): Int =
    this.uuid.compareTo(other.uuid)

  override fun toString(): String =
    this.uuid.toString()

  companion object {

    /**
     * Generate a random account ID.
     */

    fun generate(): AccountID =
      AccountID(UUID.randomUUID())
  }
}
